### PR TITLE
Test

### PR DESCRIPTION
**Pull request by Xiaoxuan1121**

### CODE REVIEW

File: tests/testthat/test-plot.R [hunk 1]
```diff
@@ -0,0 +1,3 @@
+test_that("multiplication works", {
```
Comment (dajmcdon) [inline tests/testthat/test-plot.R]: let's delete this one.

File: tests/testthat/test-sparsegl.R [hunk 2]
```diff
@@ -1,8 +1,285 @@
+library(Matrix)
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: You shouldn't need to load `Matrix` since it's a dependency of `sparsegl`.

The others need to be added to `Suggests:` in `DESCRIPTION`. Easiest is to call
```r
usethis::use_package("glmnet", type = "Suggests")
```

File: tests/testthat/test-sparsegl.R [hunk 3]
```diff
@@ -1,8 +1,285 @@
...
+library(glmnet)
+library(gglasso)
+library(SGL)
+
 test_that("multiplication works", {
+  expect_equal(2 * 2, 4)
+})
 
+test_that("x is not a matrix", {
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: 1. The name needs to be more descriptive. You're testing that "sparsegl fails when x is not a matrix".
2. You can bundle a bunch of similar tests together. So this one down to about line 83, are all "input parameter mismatches cause failures" or something like that.
3. There's no need to test the objects you create. You made `x` a vector, so we don't test that it's not a matrix, just that `sparsegl` fails.

File: tests/testthat/test-sparsegl.R [hunk 4]
```diff
@@ -1,8 +1,285 @@
...
+test_that("no penalty", {
+  
+  X <- matrix(c(1,2,3,4,5,6), nrow = 3)
+  y <- c(1.5, 3.5, 5.5)
+  res1 <- sparsegl(X, y, lambda = 0)
+  res2 <- summary(lm(y~X))
+  
+  expect_equal(res1$b0[1], res2$coefficients[1,1])
+  expect_equal(res1$beta[1], res2$coefficients[2,1])
+  expect_equal(res1$beta[2], 0)
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: Because sparsegl is a numeric method, rather than exact like `lm`, expecting these to be **equal** may not work. Better to check against a specific tolerance. By default, the tolerance in `testthat` is set to something like `1e-8`, but there's a tolerance in `sparsegl` as well. So you probably need to change the tolerance here (or at least think about it).

File: tests/testthat/test-sparsegl.R [hunk 5]
```diff
@@ -1,8 +1,285 @@
...
+
+  # try large lambda
+  expect_equal(sparsegl(X, y, lambda = 2, asparse = 1)$beta[1:2],c(0,0))
+
+  fit <- glmnet(X, y, alpha = 1)
+  lambda <- fit$lambda[c(1:2, (length(fit$lambda) - 1) :length(fit$lambda))]
+  # check the first two lambdas and last two lambdas extracted from glmnet
+
+  for (i in 1 : length(lambda)){
+    fit1 <- sparsegl(X, y, asparse = 1, lambda = lambda[i])
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: `sparsegl` should expect a vector of `lambda`

File: tests/testthat/test-sparsegl.R [hunk 6]
```diff
@@ -1,8 +1,285 @@
...
+    fit2 <- glmnet(X, y, alpha = 1)
+    print(paste("lasso: lambda = ", lambda[i]))
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: I wouldn't print output into the tests. (You can use this to check that they're doing what you want before you push, but then delete it).

File: tests/testthat/test-sparsegl.R [hunk 7]
```diff
@@ -1,8 +1,285 @@
...
+  
+  for (i in 1 : length(lambda)){
+    fit1 <- sparsegl(X, y, asparse = 1, lambda = lambda[i])
+    fit2 <- glmnet(X, y, alpha = 1)
+    print(paste("lasso: lambda = ", lambda[i]))
+    coef <- c(fit1$b0, fit1$beta[, 1])
+    print(cbind(which(coef != 0), coef[which(coef != 0)]))
+    coeff <- coef(fit2, s = lambda[i])
+    print(cbind(which(coeff != 0), coeff[which(coeff != 0)]))
+
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl.R]: We can discuss all of the code like this at our meeting.

**Review by dajmcdon — CHANGES_REQUESTED** [review]

File: tests/testthat/test-sparsegl_comparisons.R [hunk 8]
```diff
@@ -0,0 +1,169 @@
+library(Matrix)
+library(gglasso)
+library(SGL)
+usethis::use_package("glmnet", type = "Suggests")
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl_comparisons.R]: You just run this in the terminal, no need to include here.

File: tests/testthat/test-sparsegl_comparisons.R [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+requireNamespace("glmnet", quietly = TRUE)
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl_comparisons.R]: why `requireNamespace()` instead of `library()`?

File: tests/testthat/test-sparsegl_comparisons.R [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  res1 <- sparsegl(X, y, lambda = 0)
+  res2 <- summary(lm(y~X))
+  
+  coef <- c(res1$b0, res1$beta)
+  coeff <- as.numeric(res2$coefficients[, 1])
+  expect_equal(coef, coeff, tolerance = 1e-3)
+  
+})
+
+test_that("lasso penalty", {
```
Comment (dajmcdon) [inline tests/testthat/test-sparsegl_comparisons.R]: Better names: "lasso penalty...", matches something?

**Review by dajmcdon — CHANGES_REQUESTED** [review]

File: R/cv.R [hunk 11]
```diff
@@ -32,7 +32,8 @@
 #' @param foldid an optional vector of values between 1 and \code{nfold}
 #' identifying what fold each observation is in. If supplied, \code{nfold} can
 #' be missing.
-#' @param \dots other arguments that can be passed to sparsegl
+#' \code{pred.loss = "loss"}, \code{loss = "hsvm"}.
```
Comment (dajmcdon) [inline R/cv.R:35]: See above. `pred.loss` can be either `L2` or `L1`.

**Review by dajmcdon — APPROVED** [review]
